libc/src/ringbuffer.c

```diff
@@ -0,0 +1,97 @@
+#include <ringbuffer.h>
+#include <stdlib.h>
+
+int ringbuffer_init(ringbuffer_t* ref, size_t sz) {
+    ref->sz = sz;
+    ref->unr_data = 0;
+
+// alloc the needed size for this ringbuff and use kmalloc
+// if this is compiled with the proper flag
+#ifndef _KERNEL_
+    ref->data = (uint8_t*) malloc(sz);
+#else
+    ref->data = (uint8_t*) kmalloc(sz);
+#endif
+
+    if (ref->data == NULL) {
+        return -1;
+    }
+
+    ref->r_pos = 0;
+    ref->w_pos = 0;
+
+    return 0;
+}
+
+ringbuffer_t* ringbuffer_new(size_t sz) {
+    ringbuffer_t* ref = malloc(sizeof(ringbuffer_t));
+    if (ringbuffer_init(ref, sz) == -1) {
+#ifndef _KERNEL_
+        free(ref);
+#else
+        kfree(ref);
+#endif
+        ref = NULL;
+    }
+
+    return ref;
+}
+
+size_t ringbuffer_used(ringbuffer_t* ref) {
+    return ref->unr_data;
+}
+
+size_t ringbuffer_free(ringbuffer_t* ref) {
+    return ref->sz - ref->unr_data;
+}
+
+int ringbuffer_dispose(ringbuffer_t* ref) {
+#ifndef __KERNEL__
+    free(ref->data);
+    free(ref);
+#else
+    kfree(ref->data);
+    kfree(ref);
+#endif
+}
+
+int ringbuffer_write(ringbuffer_t* ref, size_t n, uint8_t* data) {
```

Comment:
This function causes buffer overflow when `n` is greater than `ref->sz` (which is a valid use case). You can check that it does with
```c
char str[] = "hello I'm about to be overwritten";
uint8_t buf[10];
uint8_t buf2[100];
ringbuffer_t* rb = ringbuffer_new(10);
rb->data = buf; // it'd do the same with the malloc'd buffer, just less visibly so
ringbuffer_write(rb, 100, buf2);
printf("str is now: %s\n", victim);
```

Reply:
I will add a guard

Reply:
a question that arose while looking more carefully at your example you expect ringbuffer_new to not allocate the buffer?

Reply:
No, that was for demonstration purposes. It's hard to show the overflow with malloc'd memory, though it would happen too, as the comment mentions.

Reply:
just an api style question I resolved it already commiting soon